File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/kafka/producers/mapping/MelosysEessiMeldingMapperA002.java
package no.nav.melosys.eessi.kafka.producers.mapping;

import java.time.LocalDate;
import no.nav.melosys.eessi.kafka.producers.model.Periode;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA002;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.SvarAnmodningUnntakBeslutning;

class MelosysEessiMeldingMapperA002 extends SvarAnmodningUnntakEessiMeldingMapper<MedlemskapA002> {

    @Override
    MedlemskapA002 hentMedlemskap(SED sed) {
        return (MedlemskapA002) sed.getMedlemskap();
    }

    @Override
    SvarAnmodningUnntakBeslutning hentBeslutning(MedlemskapA002 medlemskap) {
        String resultat = medlemskap.getUnntak().getVedtak().getResultat();

        return SvarAnmodningUnntakBeslutning.fraRinaKode(resultat);
    }

    @Override
    String hentBegrunnelse(MedlemskapA002 medlemskap) {
        return medlemskap.getUnntak().getVedtak().getBegrunnelse();
    }

    @Override
    Periode hentDelvisInnvilgetPeriode(MedlemskapA002 medlemskap) {
        if (SvarAnmodningUnntakBeslutning.fraRinaKode(medlemskap.getUnntak().getVedtak().getResultat()) == SvarAnmodningUnntakBeslutning.AVSLAG) {
            return null;
        }

        no.nav.melosys.eessi.models.sed.nav.Periode periode = medlemskap.getUnntak().getVedtak().getAnnenperiode();
        return new Periode(LocalDate.parse(periode.getFastperiode().getStartdato()), LocalDate.parse(periode.getFastperiode().getSluttdato()));
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperA002.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding

import no.nav.melosys.eessi.kafka.producers.model.Periode
import no.nav.melosys.eessi.models.DatoUtils.tilLocalDate
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA002
import no.nav.melosys.eessi.models.sed.medlemskap.impl.SvarAnmodningUnntakBeslutning
import no.nav.melosys.eessi.models.sed.medlemskap.impl.SvarAnmodningUnntakBeslutning.Companion.fraRinaKode

internal class MelosysEessiMeldingMapperA002 : SvarAnmodningUnntakEessiMeldingMapper<MedlemskapA002?>() {
    override fun hentMedlemskap(sed: SED): MedlemskapA002? = sed.medlemskap as MedlemskapA002?

    override fun hentBeslutning(medlemskap: MedlemskapA002?): SvarAnmodningUnntakBeslutning? {
        val resultat = medlemskap?.unntak!!.vedtak!!.resultat

        return fraRinaKode(resultat)
    }

    override fun hentBegrunnelse(medlemskap: MedlemskapA002?): String? = medlemskap?.unntak!!.vedtak!!.begrunnelse

    override fun hentDelvisInnvilgetPeriode(medlemskap: MedlemskapA002?): Periode? {
        if (fraRinaKode(medlemskap?.unntak!!.vedtak!!.resultat) == SvarAnmodningUnntakBeslutning.AVSLAG) {
            return null
        }

        val periode = medlemskap.unntak!!.vedtak!!.annenperiode
        return Periode(tilLocalDate(periode!!.fastperiode!!.startdato!!), tilLocalDate(periode.fastperiode!!.sluttdato!!))
    }
}
